ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+- Conventions similar to Python SDK are used here:
+  - `Client` is a general purpose interface, not limited workflow-only interactions
+  - Due to this method names are more descriptive than just `#start`, `#execute`, etc
+  - `Client#execute_workflow` is added for convenience
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a workflow_id/run_id combination using the
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if a workflow is still running after
+  #   waiting for for up to :timeout seconds.
+  # Passing timeout: nil will not wait if the working is still running
+  def result(timeout: Integer) -> Any
+
+  # Returns more information about a workflow
+  def describe() -> Temporal::WorkflowExecutionInfo
+
+  # Sends a cancellation request
+  def cancel() -> void
+
+  # Sends a query and awaits for the response
+  def query(name, *args) -> Any
+
+  # Sends a signal
+  def signal(name, *args) -> void
+
+  # Terminates a workflow
+  def terminate(reason, details: nil) -> void
+end
+```
+
+### Alternatives considered
+
+We thought about configuring the `Temporal::Client` with a namespace, similar to the Python SDK.
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
+worker_1.register_workflow(MyWorkflow) # uses 'MyWorkflow' as a name
+worker_1.register_activity(MyActivity) # uses 'MyActivity' as a name
+worker_1.start
+
+worker_2 = Temporal::Worker.new(config, 'my-namespace', 'another-task-queue')
+worker_2.register_workflow(MyWorkflow, name: 'my-workflow') # uses 'my-workflow' as a name
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
```

Comment:
I don't think this convenience method is needed. I suspect they can choose when to stop the worker based on signals themselves (as well as doing any other cleanup).

Reply:
> This means that by default it'll just stop where it is

Can you clarify?

> we're forcing users to implement graceful shutdown

Up to them. Our workers are built to crash hard or the user can call `stop()` for graceful shutdown. Obviously the latter is better because then task timeouts don't get invoked and retries don't come into play. But we definitely shouldn't assume anything about process death _inside_ the library. Process/OS concerns and when a worker is shutdown based on external stimulus are not library concerns IMO.

Also it should be noted that some SDKs provide a "graceful shutdown timeout" feature for workers that gives activities a short period to gracefully stop themselves.

In Go we do provide an opt-in shutdown-worker-on-signal feature, but it's not default on `Run`.

Reply:
This is not well-defined in the current Ruby library. This causes the user of the library to have to figure out how shutdown is impacting their worker at some later point in time, possibly only once it's causing unexpected behavior.

Whatever is done here should set those writing a worker up for success.

For example, in the Go SDK, the equivalent `.Run` function takes a channel, providing the caller with a clear mechanism for doing graceful shutdown. They can decide they don't want it by passing a `nil`, but the ambiguity is removed and they have to make a decision. Similarly, the Java SDK has a similar outcome with a different design, where there is only .start() and .stop() and you need to figure how you'd like to shutdown or just block forever if you don't care.

Reply:
Per off-PR discussion, @antstorm correct me if I'm wrong, we figure that a blocking `run()` with a `stop()` is best. That way, `run()` can return a worker runtime error and if someone wants to implement non-blocking `start()` they can just start in a new thread if they want (big bonus, we don't have to own/manage the main worker thread then, they can). So `run()` runs until `stop()` is called or fatal error. And if they just want to run `run()` until signal, they can `Signal.trap("INT") { worker.stop }` or whatever the best practice is to shutdown a thing in Ruby.

Does that seem reasonable? Or am I missing a use case?

Reply:
> Can you clarify?

Without an explicit signal handler a `SIGINT` or a `SIGTERM` will interrupt the process mid execution. This will result in unhandled workflow and activity tasks (which can be catastrophic for non-idempotent activities).

This also means that a `run` concept (blocking execution) without signal handling is not really useful since you'll have to call `stop` yourself (we're defaulting to unsafe behaviour).

The way I define the distinction is:

- `run` is for people that just want to let the worker to take over the process and forget about it
- `start`/`stop` is for people who want control and are happy to implement things manually

Basically:

```ruby
worker.run

# OR
Signal.trap('TERM') { worker.stop }
worker.start

# keep the thread alive somehow
sleep 1 unless worker.stopped?
```

We can add `stop(graceful: true)` or `stop(force: true)` (depending on the best default) to make it more configurable. But a `run` without signal handling in my opinion is a waste of a method name :)